File: backend/test/tooling/toolPipeline.test.ts
import { beforeEach, describe, expect, it } from "vitest";
import type { AccountEntitlementSnapshot, SkillDefinition, SkillExecutionContext } from "../../src/domain/types.js";
import type { ClockPort, ConfirmationPort, ConfirmationRequest, EntitlementPort, PermissionPort, UsagePort } from "../../src/tooling/ports.js";
import { SkillRegistry } from "../../src/tooling/skillRegistry.js";
import { ToolPipeline } from "../../src/tooling/toolPipeline.js";

const now = new Date("2026-08-26T00:00:00Z");
const proSnapshot: AccountEntitlementSnapshot = { accountId: "acc-1", plan: "PRO", trialExpiresAt: null, creditBalance: 100 };
const context: SkillExecutionContext = { accountId: "acc-1" };

class FakePermissionPort implements PermissionPort {
  constructor(private readonly granted: Set<string> = new Set()) {}
  async isGranted(_accountId: string, permission: string): Promise<boolean> {
    return this.granted.has(permission);
  }
}

class FakeEntitlementPort implements EntitlementPort {
  constructor(private readonly fixedSnapshot: AccountEntitlementSnapshot) {}
  async snapshot(): Promise<AccountEntitlementSnapshot> {
    return this.fixedSnapshot;
  }
}

class FakeUsagePort implements UsagePort {
  balance: number;
  charges: Array<{ skillId: string; cost: number }> = [];
  constructor(initialBalance = 100) {
    this.balance = initialBalance;
  }
  async charge(_accountId: string, cost: { value: number }, skillId: string): Promise<number> {
    this.balance -= cost.value;
    this.charges.push({ skillId, cost: cost.value });
    return this.balance;
  }
}

class FakeConfirmationPort implements ConfirmationPort {
  lastRequest: ConfirmationRequest | undefined;
  constructor(private readonly approve: boolean) {}
  async confirm(request: ConfirmationRequest): Promise<boolean> {
    this.lastRequest = request;
    return this.approve;
  }
}

const fixedClock: ClockPort = { now: () => now };

function lowRiskSkill(overrides: Partial<SkillDefinition> = {}): SkillDefinition {
  return {
    id: "web.search",
    name: "Web Search",
    description: "test skill",
    category: "WEB",
    capabilities: [],
    requiredPermissions: [],
    requiredEntitlement: "FREE",
    usageCost: { value: 0, unit: "credits" },
    riskLevel: "LOW",
    requiresConfirmation: false,
    executesOnDevice: false,
    inputSchema: { requiredFields: ["query"] },
    handler: async () => ({ kind: "success", output: { hits: 3 }, summary: "Found 3 results." }),
    ...overrides,
  };
}

function buildPipeline(
  skill: SkillDefinition,
  opts: {
    permissionPort?: FakePermissionPort;
    usagePort?: FakeUsagePort;
    confirmationPort?: FakeConfirmationPort;
    snapshot?: AccountEntitlementSnapshot;
  } = {},
) {
  const registry = new SkillRegistry();
  registry.register(skill);
  return new ToolPipeline(
    registry,
    opts.permissionPort ?? new FakePermissionPort(),
    new FakeEntitlementPort(opts.snapshot ?? proSnapshot),
    opts.usagePort ?? new FakeUsagePort(100),
    opts.confirmationPort ?? new FakeConfirmationPort(true),
    fixedClock,
  );
}

describe("ToolPipeline", () => {
  it("reports an unknown skill id as not found", async () => {
    const pipeline = buildPipeline(lowRiskSkill());
    const outcome = await pipeline.execute({ id: "1", skillId: "does.not_exist", input: { values: {} } }, context);
    expect(outcome.kind).toBe("skill_not_found");
  });

  it("fails validation before anything else runs when a required field is missing", async () => {
    const usagePort = new FakeUsagePort(100);
    const pipeline = buildPipeline(lowRiskSkill(), { usagePort });
    const outcome = await pipeline.execute({ id: "1", skillId: "web.search", input: { values: {} } }, context);
    expect(outcome).toMatchObject({ kind: "validation_failed", missingFields: ["query"] });
    expect(usagePort.charges).toHaveLength(0);
  });

  it("blocks execution on missing permission", async () => {
    const skill = lowRiskSkill({ requiredPermissions: ["CONTACTS"] });
    const pipeline = buildPipeline(skill, { permissionPort: new FakePermissionPort() });
    const outcome = await pipeline.execute({ id: "1", skillId: skill.id, input: { values: { query: "x" } } }, context);
    expect(outcome).toMatchObject({ kind: "permission_denied", missing: ["CONTACTS"] });
  });

  it("blocks and never charges on entitlement denial", async () => {
    const skill = lowRiskSkill({ requiredEntitlement: "PRO" });
    const freeSnapshot: AccountEntitlementSnapshot = { accountId: "acc-1", plan: "FREE", trialExpiresAt: null, creditBalance: 100 };
    const usagePort = new FakeUsagePort(100);
    const pipeline = buildPipeline(skill, { usagePort, snapshot: freeSnapshot });
    const outcome = await pipeline.execute({ id: "1", skillId: skill.id, input: { values: { query: "x" } } }, context);
    expect(outcome.kind).toBe("entitlement_denied");
    expect(usagePort.charges).toHaveLength(0);
  });

  it("never runs the handler when confirmation is declined", async () => {
    let handlerCalled = false;
    const skill = lowRiskSkill({
      riskLevel: "MEDIUM",
      requiresConfirmation: true,
      handler: async () => {
        handlerCalled = true;
        return { kind: "success", output: {}, summary: "should not happen" };
      },
    });
    const pipeline = buildPipeline(skill, { confirmationPort: new FakeConfirmationPort(false) });
    const outcome = await pipeline.execute({ id: "1", skillId: skill.id, input: { values: { query: "x" } } }, context);
    expect(outcome.kind).toBe("confirmation_declined");
    expect(handlerCalled).toBe(false);
  });

  it("charges the declared usage cost exactly once on success", async () => {
    const skill = lowRiskSkill({ usageCost: { value: 5, unit: "credits" } });
    const usagePort = new FakeUsagePort(100);
    const pipeline = buildPipeline(skill, { usagePort });
    const outcome = await pipeline.execute({ id: "1", skillId: skill.id, input: { values: { query: "x" } } }, context);
    expect(outcome).toMatchObject({ kind: "success", chargedCredits: 5 });
    expect(usagePort.charges).toHaveLength(1);
    expect(usagePort.balance).toBe(95);
  });

  it("fails verification and does not charge when the handler reports success with a blank summary", async () => {
    const skill = lowRiskSkill({
      usageCost: { value: 5, unit: "credits" },
      handler: async () => ({ kind: "success", output: {}, summary: "" }),
    });
    const usagePort = new FakeUsagePort(100);
    const pipeline = buildPipeline(skill, { usagePort });
    const outcome = await pipeline.execute({ id: "1", skillId: skill.id, input: { values: { query: "x" } } }, context);
    expect(outcome.kind).toBe("verification_failed");
    expect(usagePort.charges).toHaveLength(0);
  });

  it("surfaces a handler failure without charging credits", async () => {
    const skill = lowRiskSkill({
      usageCost: { value: 5, unit: "credits" },
      handler: async () => ({ kind: "failure", reason: "boom", userMessage: "Something went wrong." }),
    });
    const usagePort = new FakeUsagePort(100);
    const pipeline = buildPipeline(skill, { usagePort });
    const outcome = await pipeline.execute({ id: "1", skillId: skill.id, input: { values: { query: "x" } } }, context);
    expect(outcome).toMatchObject({ kind: "execution_failed", result: { reason: "boom" } });
    expect(usagePort.charges).toHaveLength(0);
  });
});

File: android/domain/src/test/kotlin/com/zarvismobile/domain/tooling/ToolPipelineTest.kt
package com.zarvismobile.domain.tooling

import com.zarvismobile.domain.FakeConfirmationPort
import com.zarvismobile.domain.FakeEntitlementPort
import com.zarvismobile.domain.FakePermissionPort
import com.zarvismobile.domain.FakeUsagePort
import com.zarvismobile.domain.FixedClockPort
import com.zarvismobile.domain.entity.AccountEntitlementSnapshot
import com.zarvismobile.domain.entity.EntitlementLevel
import com.zarvismobile.domain.entity.JsonSchema
import com.zarvismobile.domain.entity.PermissionType
import com.zarvismobile.domain.entity.RiskLevel
import com.zarvismobile.domain.entity.SkillCategory
import com.zarvismobile.domain.entity.SkillDefinition
import com.zarvismobile.domain.entity.SkillExecutionContext
import com.zarvismobile.domain.entity.SkillHandler
import com.zarvismobile.domain.entity.SkillInput
import com.zarvismobile.domain.entity.SkillResult
import com.zarvismobile.domain.entity.ToolCall
import com.zarvismobile.domain.entity.ToolExecutionOutcome
import com.zarvismobile.domain.entity.UsageCost
import java.time.Instant
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs
import kotlin.test.assertTrue

class ToolPipelineTest {

    private val now = Instant.parse("2026-08-26T00:00:00Z")
    private val context = SkillExecutionContext(accountId = "acc-1")
    private val proSnapshot = AccountEntitlementSnapshot("acc-1", EntitlementLevel.PRO, trialExpiresAt = null, creditBalance = 100)

    private fun registryWith(skill: SkillDefinition) = SkillRegistry().apply { register(skill) }

    private fun lowRiskSkill(
        id: String = "web.search",
        requiredPermissions: List<PermissionType> = emptyList(),
        usageCost: UsageCost = UsageCost.FREE,
        handler: SkillHandler = SkillHandler { _, _ -> SkillResult.Success(mapOf("hits" to 3), "Found 3 results.") },
    ) = SkillDefinition(
        id = id,
        name = "Web Search",
        description = "test skill",
        category = SkillCategory.WEB,
        requiredPermissions = requiredPermissions,
        requiredEntitlement = EntitlementLevel.FREE,
        usageCost = usageCost,
        riskLevel = RiskLevel.LOW,
        requiresConfirmation = false,
        inputSchema = JsonSchema(requiredFields = setOf("query")),
        handler = handler,
    )

    private fun pipeline(
        skill: SkillDefinition,
        permissionPort: FakePermissionPort = FakePermissionPort(),
        usagePort: FakeUsagePort = FakeUsagePort(100),
        confirmationPort: FakeConfirmationPort = FakeConfirmationPort(approve = true),
        snapshot: AccountEntitlementSnapshot = proSnapshot,
    ) = ToolPipeline(
        registry = registryWith(skill),
        permissionPort = permissionPort,
        entitlementPort = FakeEntitlementPort(snapshot),
        usagePort = usagePort,
        confirmationPort = confirmationPort,
        clock = FixedClockPort(now),
    )

    @Test
    fun `unknown skill id is reported as not found`() = runTest {
        val pipeline = pipeline(lowRiskSkill())
        val outcome = pipeline.execute(ToolCall(skillId = "does.not_exist", input = SkillInput()), context)
        assertIs<ToolExecutionOutcome.SkillNotFound>(outcome)
    }

    @Test
    fun `missing required input field fails validation before anything else runs`() = runTest {
        val usagePort = FakeUsagePort(100)
        val pipeline = pipeline(lowRiskSkill(), usagePort = usagePort)
        val outcome = pipeline.execute(ToolCall(skillId = "web.search", input = SkillInput(emptyMap())), context)
        val failed = assertIs<ToolExecutionOutcome.ValidationFailed>(outcome)
        assertEquals(setOf("query"), failed.missingFields)
        assertTrue(usagePort.charges.isEmpty(), "a validation failure must never charge credits")
    }

    @Test
    fun `missing permission blocks execution`() = runTest {
        val skill = lowRiskSkill(requiredPermissions = listOf(PermissionType.CONTACTS))
        val pipeline = pipeline(skill, permissionPort = FakePermissionPort(granted = emptySet()))
        val outcome = pipeline.execute(ToolCall(skillId = skill.id, input = SkillInput(mapOf("query" to "x"))), context)
        val denied = assertIs<ToolExecutionOutcome.PermissionDenied>(outcome)
        assertEquals(listOf(PermissionType.CONTACTS), denied.missing)
    }

    @Test
    fun `entitlement denial blocks execution and never charges`() = runTest {
        val skill = lowRiskSkill().copy(requiredEntitlement = EntitlementLevel.PRO)
        val freeSnapshot = AccountEntitlementSnapshot("acc-1", EntitlementLevel.FREE, trialExpiresAt = null, creditBalance = 100)
        val usagePort = FakeUsagePort(100)
        val pipeline = pipeline(skill, usagePort = usagePort, snapshot = freeSnapshot)
        val outcome = pipeline.execute(ToolCall(skillId = skill.id, input = SkillInput(mapOf("query" to "x"))), context)
        assertIs<ToolExecutionOutcome.EntitlementDenied>(outcome)
        assertTrue(usagePort.charges.isEmpty())
    }

    @Test
    fun `MEDIUM risk skill declined at confirmation never executes the handler`() = runTest {
        var handlerCalled = false
        val skill = lowRiskSkill(
            handler = SkillHandler { _, _ -> handlerCalled = true; SkillResult.Success(emptyMap(), "should not happen") },
        ).copy(riskLevel = RiskLevel.MEDIUM, requiresConfirmation = true)
        val pipeline = pipeline(skill, confirmationPort = FakeConfirmationPort(approve = false))
        val outcome = pipeline.execute(ToolCall(skillId = skill.id, input = SkillInput(mapOf("query" to "x"))), context)
        assertIs<ToolExecutionOutcome.ConfirmationDeclined>(outcome)
        assertTrue(!handlerCalled, "declining confirmation must prevent the handler from running")
    }

    @Test
    fun `successful execution charges the declared usage cost exactly once`() = runTest {
        val skill = lowRiskSkill(usageCost = UsageCost(5))
        val usagePort = FakeUsagePort(100)
        val pipeline = pipeline(skill, usagePort = usagePort)
        val outcome = pipeline.execute(ToolCall(skillId = skill.id, input = SkillInput(mapOf("query" to "x"))), context)
        val success = assertIs<ToolExecutionOutcome.Success>(outcome)
        assertEquals(5, success.chargedCredits)
        assertEquals(1, usagePort.charges.size)
        assertEquals(95, usagePort.balance)
    }

    @Test
    fun `a handler success with a blank summary fails verification and is not charged`() = runTest {
        val skill = lowRiskSkill(
            usageCost = UsageCost(5),
            handler = SkillHandler { _, _ -> SkillResult.Success(mapOf("x" to 1), "") },
        )
        val usagePort = FakeUsagePort(100)
        val pipeline = pipeline(skill, usagePort = usagePort)
        val outcome = pipeline.execute(ToolCall(skillId = skill.id, input = SkillInput(mapOf("query" to "x"))), context)
        assertIs<ToolExecutionOutcome.VerificationFailed>(outcome)
        assertTrue(usagePort.charges.isEmpty())
    }

    @Test
    fun `handler failure is surfaced without charging credits`() = runTest {
        val skill = lowRiskSkill(
            usageCost = UsageCost(5),
            handler = SkillHandler { _, _ -> SkillResult.Failure("boom", "Something went wrong.") },
        )
        val usagePort = FakeUsagePort(100)
        val pipeline = pipeline(skill, usagePort = usagePort)
        val outcome = pipeline.execute(ToolCall(skillId = skill.id, input = SkillInput(mapOf("query" to "x"))), context)
        val failed = assertIs<ToolExecutionOutcome.ExecutionFailed>(outcome)
        assertEquals("boom", failed.result.reason)
        assertTrue(usagePort.charges.isEmpty())
    }
}
